ResearchKit/ActiveTasks/ORKTowerOfHanoiStepViewController.m

```diff
@@ -0,0 +1,275 @@
+
+- (NSMutableArray *)moves {
+    if (_moves) {
+        return _moves;
+    }
+    _moves = [NSMutableArray array];
+    return _moves;
+}
+
+- (NSDateComponentsFormatter *)dateComponentsFormatter {
+    if (_dateComponentsFormatter) {
+        return _dateComponentsFormatter;
+    }
+    _dateComponentsFormatter = [NSDateComponentsFormatter new];
+    _dateComponentsFormatter.unitsStyle = NSDateComponentsFormatterUnitsStylePositional;
+    _dateComponentsFormatter.zeroFormattingBehavior = NSDateComponentsFormatterZeroFormattingBehaviorPad;
+    _dateComponentsFormatter.allowedUnits =  NSCalendarUnitMinute | NSCalendarUnitSecond;
+    return _dateComponentsFormatter;
+}
+
+- (NSUInteger) numberOfDisks {
+    return ((ORKTowerOfHanoiStep *)self.step).numberOfDisks;
+}
+
+- (void)updateTitleText {
+    NSString *moves = ORKLocalizedStringFromNumber(@(self.moves.count));
+    NSString *time = [self.dateComponentsFormatter stringFromTimeInterval:_secondsElapsed];
+    NSString *text = [NSString stringWithFormat:ORKLocalizedString(@"TOWER_OF_HANOI_TASK_ACTIVE_STEP_PROGRESS_TEXT", nil), moves, time];
+    [self.activeStepView updateTitle:nil text:text];
+}
+
+- (void)reloadData {
+    for (ORKTowerOfHanoiTowerView *towerView in _towerViews) {
+        towerView.highlighted = _selectedIndex != nil && [_towerViews indexOfObject:towerView] == _selectedIndex.integerValue;
+        [towerView reloadData];
+    }
+}
+
+- (BOOL)puzzleIsSolved {
+    return ((ORKTowerOfHanoiTower *)_towers.lastObject).disks.count == [self numberOfDisks];
+}
+
+- (void)evaluatePuzzle {
+    if ([self puzzleIsSolved]) {
+        [self finish];
+    }
+}
+
+- (void)setupTowers {
+    NSMutableArray *diskStack = [NSMutableArray array];
+    for (NSInteger disk = [self numberOfDisks] ; disk > 0 ; disk--) {
+        [diskStack addObject: @(disk)];
+    }
+    _towers = [@[[[ORKTowerOfHanoiTower alloc] initWithDisks:diskStack], [ORKTowerOfHanoiTower emptyTower], [ORKTowerOfHanoiTower emptyTower]] mutableCopy];
+}
+
+- (void)setupTowerViews {
+    NSMutableArray *towerViews = [NSMutableArray array];
+    for (NSInteger index = 0 ; index < 3 ; index++) {
+        ORKTowerOfHanoiTowerView *towerView = [[ORKTowerOfHanoiTowerView alloc] initWithFrame:CGRectZero maximumNumberOfDisks:[self numberOfDisks]];
+        towerView.delegate = self;
+        towerView.dataSource = self;
+        towerView.targeted = (index == kNumberOfTowers - 1);
+        [towerViews addObject:towerView];
+        [towerView setTranslatesAutoresizingMaskIntoConstraints:NO];
+        [_towerOfHanoiCustomView addSubview:towerView];
+    }
+    _towerViews = towerViews;
+}
+
+- (void)transferDiskFromTowerAtIndex:(NSInteger)donorTowerIndex toTowerAtIndex:(NSInteger)recipientTowerIndex {
+    ORKTowerOfHanoiTower *donorTower = _towers[donorTowerIndex];
+    ORKTowerOfHanoiTower *recipientTower = _towers[recipientTowerIndex];
+    if (donorTower.disks.count > 0 && [recipientTower recieveDiskFrom:donorTower]) {
```

Comment:
Would it be nicer to move `donorTower.disks.count > 0` into `recieveDiskFrom:` ?